server/src/domain/database/database.service.ts

```diff
@@ -0,0 +1,85 @@
+import { DatabaseExtension, IDatabaseRepository } from '../repositories';
+
+@Injectable()
+export class DatabaseService {
+  private logger = new ImmichLogger(DatabaseService.name);
+  private _minVectorsVersion: Version = new Version(0, 1, 1);
+  private _maxVectorsVersion: Version = new Version(0, 1, 11);
+
+  constructor(@Inject(IDatabaseRepository) private databaseRepository: IDatabaseRepository) {}
+
+  async init() {
+    await this.createVectors();
+    await this.assertVectors();
+    await this.databaseRepository.runMigrations();
+  }
+
+  private async assertVectors() {
+    const version = await this.databaseRepository.getExtensionVersion(DatabaseExtension.VECTORS);
+    if (version == null) {
+      throw new Error('Unexpected: The pgvecto.rs extension is not installed.');
+    }
+
+    const image = await this.getVectorsImage();
+    const suggestion = image ? `, such as with the docker image '${image}'` : '';
+
+    if (version.isEqual(new Version(0, 0, 0))) {
+      throw new Error(
+        `The pgvecto.rs extension version is ${version}, which means it is a nightly release.` +
+          `Please run 'DROP EXTENSION IF EXISTS vectors' and switch to a release version${suggestion}.`,
+      );
+    }
+
+    if (version.isNewerThan(this.maxVectorsVersion)) {
+      throw new Error(`
+        The pgvecto.rs extension version is ${version} instead of ${this.maxVectorsVersion}.
+        Please run 'DROP EXTENSION IF EXISTS vectors' and switch to ${this.maxVectorsVersion}${suggestion}.`);
+    }
+
+    if (version.isOlderThan(this.minVectorsVersion)) {
+      throw new Error(`
+        The pgvecto.rs extension version is ${version}, which is older than the minimum supported version ${this.minVectorsVersion}.
+        Please upgrade to this version or later${suggestion}.`);
+    }
+  }
+
+  private async createVectors() {
+    await this.databaseRepository.createExtension(DatabaseExtension.VECTORS).catch(async (err: QueryFailedError) => {
+      const image = await this.getVectorsImage();
+      this.logger.fatal(`
+        Failed to create pgvecto.rs extension.
+        If you have not updated your Postgres instance to a docker image that supports pgvecto.rs (such as '${image}'), please do so.
+        See the v1.91.0 release notes for more info: https://github.com/immich-app/immich/releases/tag/v1.91.0'
+      `);
+      throw err;
+    });
+  }
+
+  private async getVectorsImage() {
+    const { major } = await this.databaseRepository.getPostgresVersion();
+    if (![14, 15, 16].includes(major)) {
+      return null;
+    }
+    return `tensorchord/pgvecto-rs:pg${major}-v${this.maxVectorsVersion}`;
+  }
+
+  get minVectorsVersion() {
+    return this._minVectorsVersion;
+  }
+
+  set minVectorsVersion(version: Version) {
+    this._minVectorsVersion = version;
+  }
+
+  get maxVectorsVersion() {
+    return this._maxVectorsVersion;
+  }
+
+  set maxVectorsVersion(version: Version) {
+    this._maxVectorsVersion = version;
+  }
```

Comment:
What is the purpose of these getters and setters vs just public properties on the class? Also, do we need this to ever be anything but hard coded/read only values?

Reply:
They don't really serve any point. I originally added a getter because `expectedVectorsVersions[expectedVectorsVersions.length - 1]` was too long, but then I changed that to separate min/max properties anyway. They're hardcoded in practice, but I made them set-able for the tests.